bibblio.py

```diff
@@ -0,0 +1,377 @@
+        except Exception as e:
+            return CoverageFailure(
+                identifier, str(e), data_source=self.output_source,
+                transient=True
+            )
+
+        content_item_id = result.get('contentItemId')
+        bibblio_identifier, _is_new = Identifier.for_foreign_id(
+            self._db, Identifier.BIBBLIO_CONTENT_ITEM_ID, content_item_id
+        )
+
+        identifier.equivalent_to(
+            self.output_source, bibblio_identifier, 1
+        )
+        return identifier
+
+    def add_coverage_record_for(self, item):
+        """Adds an appropriate CoverageRecord for this particular
+        identifier and any equivalent identifiers.
+
+        Because Bibblio is text-based recommendation source, it doesn't
+        make sense to represent the same work multiple times.
+        """
+        equivalent_identifier_ids = item.equivalent_identifier_ids()[item.id]
+
+        equivalent_identifiers = self._db.query(Identifier).filter(
+            Identifier.id.in_(equivalent_identifier_ids)
+        )
+        for identifier in equivalent_identifiers:
+            super(BibblioCoverageProvider, self).add_coverage_record_for(identifier)
+
+    def content_item_from_identifier(self, identifier):
+        edition = identifier.primarily_identifies[0]
+
+        name = edition.title + ' by ' + edition.author
+        url = self.edition_permalink(edition)
+        text, data_source = self.get_full_text(edition)
+        provider = dict(name=data_source.name)
+
+        content_item = dict(
+            name=name, url=url, text=text, provider=provider
+        )
+        if self.catalogue_id:
+            content_item['catalogueId'] = self.catalogue_id
+
+        return content_item
+
+    @classmethod
+    def edition_permalink(cls, edition):
+        base_url = Configuration.integration_url(
+            Configuration.CONTENT_SERVER_INTEGRATION, required=True
+        )
+        scheme, host = urlparse(base_url)[0:2]
+        base_url = '://'.join([scheme, host])
+
+        urn = edition.primary_identifier.urn
+        permalink = '%s/lookup?urn=%s' % (base_url, urn)
+        return permalink
+
+    def get_full_text(self, edition_or_identifier):
+        identifier = edition_or_identifier
+        if not isinstance(edition_or_identifier, Identifier):
+            identifier = edition_or_identifier.primary_identifier
+
+        representations = self._db.query(Representation)\
+            .join(Representation.resource).join(Resource.links)\
+            .join(Hyperlink.identifier).filter(
+                Identifier.id==identifier.id,
+                Hyperlink.rel==Hyperlink.OPEN_ACCESS_DOWNLOAD)\
+            .options(eagerload(Representation.resource))
+
+        text_representation = representations.filter(
+            Representation.media_type.in_(self.TEXT_MEDIA_TYPES),
+            Representation.content.isnot(None))\
+            .limit(1).all()
```

Comment:
I think it would be simpler to go through LicensePool -> LicensePoolDeliveryMechanism -> Resource. Unless I'm missing something about looking for the text in formats that don't necessarily become LicensePoolDeliveryMechanisms.